Add AWS X-Ray Exporter

kbrockhoff opened the pull request:
Documentation is covered in README in awsxrayexporter directory. conn.go and X-ray_client.go were adapted from https://github.com/aws/aws-xray-daemon

kbrockhoff committed:
initial def of aws xray exporter

kbrockhoff committed:
initial def of aws xray exporter

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of xray data structures and converters

kbrockhoff committed:
initial dev of export handler

kbrockhoff committed:
initial dev of export handler

kbrockhoff committed:
Merge remote-tracking branch 'upstream/master'

kbrockhoff committed:
fix formatting and lint errors

kbrockhoff committed:
fix formatting and lint errors

kbrockhoff committed:
initial dev of export handler

kbrockhoff committed:
initial dev of export handler

kbrockhoff committed:
added to component list

kbrockhoff committed:
added to component list

owais commented on exporter/awsxrayexporter/translator/cause.go:
```diff
@@ -0,0 +1,111 @@
+	Label string `json:"label,omitempty"`
+}
+
+func makeCause(status *tracepb.Status, attributes map[string]string) (bool, bool, map[string]string, *CauseData) {
```
would be nice to document what the return variables mean or even use named returns if simple to do.

kbrockhoff replied:
Fixed

owais commented on exporter/awsxrayexporter/translator/writer_pool.go:
```diff
@@ -0,0 +1,69 @@
+		w.buffer.Reset()
+		writers.Put(w)
+	}
+}
```
Might be better to be able to initialize multiple pools and add the borrow/return method to the pool itself instead of one global pool. Would make it easier to test among other things.

kbrockhoff replied:
Fixed

owais reviewed (commented):
Did a quick first pass and added some comments. I see a lot of debug/info statements which suggests this is probably still WIP and not ready for review? Feel free to ignore comments for now if that is the case but please add WIP to the title.

tigrannajaryan commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,276 @@
+			switch aerr.Code() {
+			case sts.ErrCodeRegionDisabledException:
+				logger.Error("Region : %v - %v", zap.String("region", region), zap.String("error", aerr.Error()))
+				logger.Info("Credentials for provided RoleARN will be fetched from STS primary region endpoint instead of regional endpoint.")
```
Is there a need for both Error and Info? Would it be better to just do one Error logging?

kbrockhoff replied:
Fixed

tigrannajaryan commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,276 @@
+	result, serr := session.NewSession()
+	if serr != nil {
+		logger.Error("Error in creating session object : %v\n.", zap.Error(serr))
+		os.Exit(1)
```
We should not be exiting on errors. Instead log if necessary and propagate the error to the caller.

Generally as a rule of thumb: Collector is a long-running process that needs to be autonomously robust. Exiting the process on an error will result in losing all data queued inside collector (if this function is called while the Collector is running).

kbrockhoff replied:
Fixed

kbrockhoff committed:
fix issues raised during code review

kbrockhoff committed:
fix issues raised during code review

kbrockhoff committed:
switch user attribute name to constant

kbrockhoff committed:
switch user attribute name to constant

kbrockhoff commented on cmd/otelcontribcol/components.go:
```diff
@@ -21,6 +21,7 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector/oterr"
 	"github.com/open-telemetry/opentelemetry-collector/receiver"
 
+	//"github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsxrayexporter"
```
Fixed

kbrockhoff commented on exporter/awsxrayexporter/config.go:
```diff
@@ -0,0 +1,42 @@
+	// Enable or disable TLS certificate verification.
+	NoVerifySSL bool `mapstructure:"no_verify_ssl"`
+	// Upload segments to AWS X-Ray through a proxy.
+	ProxyAddress string `mapstructure:"proxy_address"`
```
This was copied from AWS X-Ray Daemon. Given the long-term heavy usage of the daemon where they have long ago fixed edge case issues, I think we should leave as is.

kbrockhoff commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,276 @@
+)
+
+// getNewHTTPClient returns new HTTP client instance with provided configuration.
+func getNewHTTPClient(logger *zap.Logger, maxIdle int, requestTimeout int, noVerify bool,
```
Fixed

kbrockhoff commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,276 @@
+	} else {
+		finalProxyAddress = ""
+	}
+	return finalProxyAddress
```
This was copied from AWS X-Ray Daemon. Given the long-term heavy usage of the daemon where they have long ago fixed edge case issues, I think we should leave as is.

kbrockhoff commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,276 @@
+		proxyURL, err = url.Parse(finalProxyAddress)
+		if err != nil {
+			//log.Errorf("Bad proxy URL: %v", err)
+			os.Exit(1)
```
Fixed

kbrockhoff commented on exporter/awsxrayexporter/testdata/config.yaml:
```diff
@@ -0,0 +1,23 @@
+    region: eu-west-1
+    resource_arn: "arn:aws:ec2:us-east1:123456789:instance/i-293hiuhe0u"
+    role_arn: "arn:aws:iam::123456789:role/monitoring-EKS-NodeInstanceRole"
+    user_attribute: "user.id"
```
All configuration values are used by AWS client so really there is only one section.

kbrockhoff commented on exporter/awsxrayexporter/translator/aws.go:
```diff
@@ -0,0 +1,179 @@
+	AwsRegionAttribute    = "aws.region"
+	AwsRequestIDAttribute = "aws.request_id"
+	AwsQueueURLAttribute  = "aws.queue_url"
+	AwsTableNameAttribute = "aws.table_name"
```
Fixed

kbrockhoff commented on exporter/awsxrayexporter/translator/aws.go:
```diff
@@ -0,0 +1,179 @@
+	if origin == "" {
+		return filtered, awsData
+	}
+	awsData = &AWSData{
```
Fixed

kbrockhoff commented on cmd/otelcontribcol/components.go:
```diff
@@ -41,7 +42,10 @@ func components() (config.Factories, error) {
 	}
 
-	exporters := []exporter.Factory{&stackdriverexporter.Factory{}}
+	exporters := []exporter.Factory{
```
Fixed

kbrockhoff commented on exporter/awsxrayexporter/awsxray.go:
```diff
@@ -0,0 +1,76 @@
+	return exporterhelper.NewTraceExporter(
+		config,
+		func(ctx context.Context, td consumerdata.TraceData) (int, error) {
+			logger.Info("TraceExporter", typeLog, nameLog, zap.Int("#spans", len(td.Spans)))
```
Fixed

kbrockhoff committed:
Merge remote-tracking branch 'upstream/master'

kbrockhoff commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,276 @@
+
+// getPartition return AWS Partition for the provided region.
+func getPartition(region string) string {
+	p, _ := endpoints.PartitionForRegion(endpoints.DefaultPartitions(), region)
```
Second return is bool not error. p can never be null.

func PartitionForRegion(ps []Partition, regionID string) (Partition, bool) {
	for _, p := range ps {
		if _, ok := p.p.Regions[regionID]; ok || p.p.RegionRegex.MatchString(regionID) {
			return p, true
		}
	}

	return Partition{}, false
}

kbrockhoff commented on exporter/awsxrayexporter/config.go:
```diff
@@ -0,0 +1,42 @@
+	// X-Ray service endpoint to which the collector sends segment documents.
+	Endpoint string `mapstructure:"endpoint"`
+	// Number of seconds before timing out a request.
+	RequestTimeout int `mapstructure:"request_timeout"`
```
Fixed

kbrockhoff replied:
Fixed

kbrockhoff committed:
fixed additional code review issues

kbrockhoff committed:
fixed additional code review issues

kbrockhoff commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,276 @@
+	config := &aws.Config{
+		Region:                 aws.String(awsRegion),
+		DisableParamValidation: aws.Bool(true),
+		MaxRetries:             aws.Int(2),
```
Fixed

kbrockhoff commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,276 @@
+				logger.Error("Region : %v - %v", zap.String("region", region), zap.String("error", aerr.Error()))
+				logger.Info("Credentials for provided RoleARN will be fetched from STS primary region endpoint instead of regional endpoint.")
+				stsCred = getSTSCredsFromPrimaryRegionEndpoint(logger, t, roleArn, region)
+			}
```
Not necessary because of the way STS is implemented.

tigrannajaryan commented on exporter/awsxrayexporter/translator/http.go:
```diff
@@ -0,0 +1,290 @@
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+// OpenTelemetry Semantic Convention attribute names for HTTP and gRPC related attributes
```
Issue to work on: https://github.com/open-telemetry/opentelemetry-collector/issues/435

kbrockhoff replied:
moved to collector

tigrannajaryan commented on exporter/awsxrayexporter/translator/http.go:
```diff
@@ -0,0 +1,290 @@
+	ContentLength int64 `json:"content_length,omitempty"`
+}
+
+func convertToHTTPStatusCode(code int32) int64 {
```
There is already a mapping from HTTP to gRPC in `github.com/open-telemetry/opentelemetry-collector/translator/trace` package. It would be best to add the corresponding reverse translation that is done by this function to that same package for consistency.
Will you be able to create a separate PR for that at https://github.com/open-telemetry/opentelemetry-collector/ ?

kbrockhoff replied:
Moved to collector

tigrannajaryan commented on exporter/awsxrayexporter/translator/segment.go:
```diff
@@ -0,0 +1,367 @@
+// OpenTelemetry Semantic Convention values for general Span attribute names.
+const (
+	ComponentAttribute   = "component"
+	PeerAddressAttribute = "peer.address"
```
Same for this, would be best to move all OpenTelemetry semantic convention constants to core repo.

kbrockhoff replied:
Moved to collector

kbrockhoff committed:
Merge remote-tracking branch 'upstream/master'

kbrockhoff committed:
fixed additional code review issues

kbrockhoff committed:
fixed additional code review issues

kbrockhoff committed:
temporarily change package name

kbrockhoff committed:
temporarily change package name

kbrockhoff committed:
temporarily change package name

kbrockhoff committed:
temporarily change package name

kbrockhoff committed:
revert temporarily change package name

kbrockhoff committed:
revert temporarily change package name

kbrockhoff committed:
fixed additional code review issues

kbrockhoff committed:
fixed additional code review issues

kbrockhoff committed:
Merge remote-tracking branch 'upstream/master'

kbrockhoff committed:
switched to constants defined in collector

kbrockhoff committed:
switched to constants defined in collector

kbrockhoff committed:
Merge remote-tracking branch 'upstream/master'

kbrockhoff committed:
switched to status conversion functions defined in collector

kbrockhoff committed:
switched to status conversion functions defined in collector

kbrockhoff commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,300 @@
+		logger.Error("unable to obtain proxy URL", zap.Error(err))
+		return nil, nil, err
+	}
+	regionEnv := os.Getenv("AWS_REGION")
```
This how aws-xray-daemon handles it. Since this is the battle-tested approach, I would prefer to leave as is.

tigrannajaryan commented:
@kbrockhoff please resolve merge conflicts.

tigrannajaryan commented on exporter/awsxrayexporter/doc.go:
```diff
@@ -0,0 +1,17 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package awsxrayexporter implements an Otel Collector exporter that sends trace data to
```
```suggestion
// Package awsxrayexporter implements an OpenTelemetry Collector exporter that sends trace data to
```

kbrockhoff replied:
Fixed

tigrannajaryan commented on exporter/awsxrayexporter/translator/segment.go:
```diff
@@ -0,0 +1,323 @@
+	writers = newWriterPool(2048)
+)
+
+// MakeSegmentDocumentString converts an Otel Span to an X-Ray Segment and then serialzies to JSON
```
This is actually an OpenCensus Span, not Otel Span. Same for function below.

kbrockhoff replied:
Fixed

tigrannajaryan commented on exporter/awsxrayexporter/go.mod:
```diff
@@ -0,0 +1,172 @@
+go 1.12
+
+require (
+	cloud.google.com/go/datastore v1.0.0 // indirect
```
Can you please make sure `go mod tidy` was run so that there are no unneeded dependencies?

kbrockhoff replied:
Ran tidy

tigrannajaryan commented on exporter/awsxrayexporter/translator/cause.go:
```diff
@@ -0,0 +1,115 @@
+	tracepb "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+)
+
+// OpenTelemetry Semantic Convention attribute names for error/fault/exception related attributes
```
Is this comment valid? I do not see these anywhere in OpenTelemetry spec.

kbrockhoff replied:
Removed for now

kbrockhoff committed:
Merge remote-tracking branch 'upstream/master'

kbrockhoff committed:
fix latest code review issues

kbrockhoff committed:
fix latest code review issues

tigrannajaryan commented:
@kbrockhoff can you please reply to comments which you addressed? This is a very big PR and it is difficult to track what is already done and what not.

kbrockhoff commented:
All comments have been addressed

kbrockhoff commented on exporter/awsxrayexporter/config_test.go:
```diff
@@ -0,0 +1,61 @@
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
```
Fixed

kbrockhoff commented on exporter/awsxrayexporter/config.go:
```diff
@@ -0,0 +1,42 @@
+type Config struct {
+	configmodels.ExporterSettings `mapstructure:",squash"` // squash ensures fields are correctly decoded in embedded struct.
+	// Maximum number of concurrent calls to AWS X-Ray to upload documents.
+	Concurrency int `mapstructure:"num_workers"`
```
Switched to same name

kbrockhoff commented on exporter/awsxrayexporter/conn.go:
```diff
@@ -0,0 +1,300 @@
+	regionEnv := os.Getenv("AWS_REGION")
+	if cfg.Region == "" && regionEnv != "" {
+		awsRegion = regionEnv
+		logger.Debug("Fetch region %v from environment variables", zap.String("region", awsRegion))
```
Fixed

kbrockhoff committed:
Merge remote-tracking branch 'upstream/master'

kbrockhoff committed:
fix new static check issues

kbrockhoff committed:
fix new static check issues

kbrockhoff committed:
Merge remote-tracking branch 'upstream/master'

tigrannajaryan reviewed (approved):
LGTM.
This is a big PR, I would like at least another pair of eyes to have a look.

rghetia reviewed (commented):
I ran the test on this PR and it failed.
```
➜  awsxrayexporter git:(otelsvcct_pr_go_41) ✗ pwd
/Users/rahulpa/rghetia/go/src/github.com/open-telemetry/opentelemetry-service-contrib/exporter/awsxrayexporter
➜  awsxrayexporter git:(otelsvcct_pr_go_41) ✗ go test -v ./
=== RUN   TestTraceExport
--- FAIL: TestTraceExport (5.31s)
    awsxray_test.go:42:
        	Error Trace:	awsxray_test.go:42
        	Error:      	Expected nil, but got: &awserr.baseError{code:"NoCredentialProviders", message:"no valid providers in chain. Deprecated.\n\tFor verbose messaging see aws.Config.CredentialsChainVerboseErrors", errs:[]error(nil)}
        	Test:       	TestTraceExport
=== RUN   TestLoadConfig
--- PASS: TestLoadConfig (0.01s)
=== RUN   TestEC2Session
--- PASS: TestEC2Session (0.00s)
=== RUN   TestRegionEnv
--- PASS: TestRegionEnv (0.00s)
=== RUN   TestCreateDefaultConfig
--- PASS: TestCreateDefaultConfig (0.00s)
=== RUN   TestCreateTraceExporter
--- PASS: TestCreateTraceExporter (0.33s)
=== RUN   TestCreateMetricsExporter
--- PASS: TestCreateMetricsExporter (0.00s)
FAIL
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsxrayexporter	5.868s
FAIL
```

Circle-ci doesn't seem to run the test on this exporter. See the log captured from the latest [run](https://circleci.com/api/v1.1/project/github/open-telemetry/opentelemetry-collector-contrib/195/output/104/0?file=true&allocation-id=5df12b775523f5574cee3e2f-0-build%2FF2189CE)
```
go test -race -timeout 30s github.com/open-telemetry/opentelemetry-collector-contrib/cmd/otelcontribcol github.com/open-telemetry/opentelemetry-collector-contrib/exporter/azuremonitorexporter github.com/open-telemetry/opentelemetry-collector-contrib/exporter/sapmexporter github.com/open-telemetry/opentelemetry-collector-contrib/exporter/signalfxexporter github.com/open-telemetry/opentelemetry-collector-contrib/exporter/stackdriverexporter github.com/open-telemetry/opentelemetry-collector-contrib/internal/version github.com/open-telemetry/opentelemetry-collector-contrib/receiver/collectdreceiver github.com/open-telemetry/opentelemetry-collector-contrib/receiver/sapmreceiver github.com/open-telemetry/opentelemetry-collector-contrib/receiver/zipkinscribereceiver github.com/open-telemetry/opentelemetry-collector-contrib/testbed/tests
```

kbrockhoff committed:
fix test that breaks if no valid aws session available

kbrockhoff committed:
fix test that breaks if no valid aws session available

tigrannajaryan committed:
Update Collector Core dependency to latest `master` (#61)

- Updated go.mod and testbed/go.mod to point to latest `master` commit
  for github.com/open-telemetry/opentelemetry-collector dependencies and
  fixed the code as needed.
- Run `go mod tidy` on both go.mod files.

rghetia commented on exporter/awsxrayexporter/translator/segment.go:
```diff
@@ -0,0 +1,320 @@
+}
+
+// NewTraceID generates a new valid X-Ray TraceID
+func NewTraceID() []byte {
```
This is only used in test. Does this need to be exported?

kbrockhoff replied:
Fixed

rghetia commented on exporter/awsxrayexporter/translator/segment.go:
```diff
@@ -0,0 +1,320 @@
+}
+
+// NewSegmentID generates a new valid X-Ray SegmentID
+func NewSegmentID() []byte {
```
same as above.

kbrockhoff replied:
Fixed

rghetia commented on exporter/awsxrayexporter/translator/segment.go:
```diff
@@ -0,0 +1,320 @@
+// fixSegmentName removes any invalid characters from the span name.  AWS X-Ray defines
+// the list of valid characters here:
+// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html
+func fixSegmentName(name string) string {
```
Add a test for SpanName with Invalid Characters. OR add a TODO to add the test later.

kbrockhoff replied:
Added test

rghetia commented on exporter/awsxrayexporter/translator/segment.go:
```diff
@@ -0,0 +1,320 @@
+// fixAnnotationKey removes any invalid characters from the annotaiton key.  AWS X-Ray defines
+// the list of valid characters here:
+// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html
+func fixAnnotationKey(key string) string {
```
same as above.

kbrockhoff replied:
Added test

rghetia commented on exporter/awsxrayexporter/translator/segment.go:
```diff
@@ -0,0 +1,320 @@
+	return float64(t.UnixNano()) / 1e9
+}
+
+func mergeAnnotations(dest map[string]interface{}, src map[string]string) {
```
this method is simply sanitizing Annotation keys. mergeAnnotation is misleading name.

Also, it is possible that two keys are same after sanitization. Although the scenario is a corner case, is the expectation that the last value is exported?

kbrockhoff replied:
Fixed

rghetia commented on exporter/awsxrayexporter/translator/segment.go:
```diff
@@ -0,0 +1,320 @@
+func makeAnnotations(attributes map[string]string) (string, map[string]interface{}) {
+	var (
+		result = map[string]interface{}{}
+		user   string
```
user is never set. why return the user?

kbrockhoff replied:
Fixed by removing

rghetia reviewed (commented):
Test is not running for awsxrayexporter because the package doesn't exist on the github yet. Ideally, bare bone exporter module should have been added first and then this PR.

Apart from that few more comments.

owais committed:
Ported kinesis exporter from Omnition (#60)

Porting the existing kinesis exporter from Omnition's Otel distribution
to contrib.

Porting from: https://github.com/Omnition/omnition-opentelemetry-collector/tree/master/exporter/kinesis

kbrockhoff committed:
fix pull request issues

kbrockhoff committed:
fix pull request issues

owais committed:
Updated SAPM exporter dependencies (#59)

tigrannajaryan committed:
Add E2E test for SAPM Receiver (#63)

- Added DataSender for SAPM protocol.
- Added SAPM protocol receiver test to TestTrace10K scenario.

rghetia reviewed (approved):
LGTM. However, it should be merged after #65 is merged. This will ensure that test for this new module runs in the CI build.

owais committed:
Updated component dependencies (#64)

rghetia committed:
Fix build for go 1.13 (#65)

rghetia commented:
@kbrockhoff #65 is merged. Can you please rebase this?

tigrannajaryan committed:
Add E2E test for SAPM Exporter (#66)

SAPM trace test was using SAPM Receiver and Jaeger Exporter.
Now SAPM is used for both ends.

kbrockhoff committed:
Merge branch 'master' of github.com:kbrockhoff/opentelemetry-collector-contrib
